python/plugins/processing/algs/qgis/SplitWithLines.py

```diff
@@ -81,107 +81,111 @@ def processAlgorithm(self, progress):
+                                if splitterPList == None:
+                                    splitterPList = vector.extractPoints(splitGeom)
+
+                                try:
+                                    result, newGeometries, topoTestPoints = inGeom.splitGeometry(splitterPList, False)
+                                except:
+                                    ProcessingLog.addToLog(ProcessingLog.LOG_WARNING,
+                                                           self.tr('Geometry exception while splitting'))
+                                    result = 1
+
+                                # splitGeometry: If there are several intersections
+                                # between geometry and splitLine, only the first one is considered.
+                                if result == 0:  # split occurred
+                                    if inPoints == vector.extractPoints(inGeom):
+                                        # bug in splitGeometry: sometimes it returns 0 but
+                                        # the geometry is unchanged
                                         outGeoms.append(inGeom)
+                                    else:
+                                        inGeoms.append(inGeom)
+
+                                        for aNewGeom in newGeometries:
+                                            inGeoms.append(aNewGeom)
                                 else:
                                     outGeoms.append(inGeom)
+                            else:
+                                outGeoms.append(inGeom)
 
-                            inGeoms = outGeoms
+                        inGeoms = outGeoms
 
-                for aGeom in inGeoms:
-                    passed = True
+            parts = []
 
-                    if QgsWkbTypes.geometryType( aGeom.wkbType() )  == QgsWkbTypes.LineGeometry \
-                            and not QgsWkbTypes.isMultiType(aGeom.wkbType()):
-                        passed = len(aGeom.asPolyline()) > 2
+            for aGeom in inGeoms:
+                passed = True
 
-                        if not passed:
-                            passed = (len(aGeom.asPolyline()) == 2 and
-                                      aGeom.asPolyline()[0] != aGeom.asPolyline()[1])
-                            # sometimes splitting results in lines of zero length
+                if QgsWkbTypes.geometryType( aGeom.wkbType() )  == QgsWkbTypes.LineGeometry:
+                    firstVertex = aGeom.vertexAt(0)
+                    secondVertex  = aGeom.vertexAt(1)
+                    thirdVertex = aGeom.vertexAt(2)
+                    passed = thirdVertex != QgsPoint(0, 0) # more than 2 vertices
 
-                    if passed:
-                        outFeat.setGeometry(aGeom)
-                        writer.addFeature(outFeat)
+                    if not passed:
+                        passed = secondVertex != QgsPoint(0, 0) # has 2 vertices
```

Comment:
Actually.... this whole block is rather confusingly written. I'd suggest changing it to something like:

```
num_points = aGeom.geometry().numPoints()
if num_points > 2:
    passed = True
elif num_points == 2:
    passed = not aGeom.geometry().isClosed() # tests if vertex 0 = vertex 1
else:
    # not enough vertices to be a valid line
    passed = False
```